### ISSUE TITLE
OpenEO deploy with gatekeeper causes issues

### ISSUE THREAD

**Issue by JeroenVerstraelen** [Open-EO/openeo-geopyspark-driver#328]
When deploying openEO it is sometimes required to manually disable and then re-enable the gatekeeper service.

**Comment by jdries** [comment]
The error is caused by tsservice creating a spark context before the openEO spark context is set up. This can happen because the tsservice is registered with gatekeeper before openEO is online, so monitoring checks start hitting tsservice and trigger context creation.

Note that we also want to get rid of gatekeeper entirely and would rather switch to traefik, like we do for openEO, which seems quite feasible.

**Comment by bossie** [comment]
Notes:

Envar `OPENEO_TRAEFIK_SERVER_ID` is ultimately used to identify a particular (OpenEO) server in Traefik. When a server registers itself in Traefik, it will essentially just update the  `url` znode for this ID.

 A server znode is also never deleted (like when it crashes or someone kills it); rather, Traefik will detect a failing health check and remove the server from the load-balanced set.

**Comment by bossie** [comment]
The way the Terrascope OpenEO deployment currently works is:

1. instantiate a SparkContext
2. determine a random port on this host
3. run gunicorn/Flask on that port
4. (asynchronously) register (host and) port in Traefik/Zookeeper

In the same vain, we can start Jetty/tsservice synchronously (instead of asynchronously with a `RunTimeSeriesServiceListener`) and register it in Traefik/ZK as well:

- create a Traefik HTTP service e.g. "tsservice-prod" in Traefik backed by Jetty/tsservice servers on epod-openeo-1 and epod-openeo-2
- create a Traefik router with a rule that maps tsservice requests to said service "tsservice-prod"
- point upstream reverse proxies to traefik0[12].vgt.vito.be instead of what looks like nginx on epod-openeo-1 (TBD)

**Comment by bossie** [comment]
Dropped `RunTimeSeriesServiceListener` in favor of a Jetty/tsservice that:

- is only started after creating the OpenEO SparkContext,
- is registered in Traefik ("tsservice-openeo-prod" and "tsservice-openeo-dev").

This makes the integration tests run but now requests have to go to Traefik instead.

**Comment by bossie** [comment]
In the gatekeeper case, the distinction between dev and prod clusters of tsservice was based on the fact that there are multiple nginx/gatekeeper setups: the ones on epod-tsservice-[12].vgt.vito.be considered the "tsservice-openeo-prod" branch in Zookeeper, while the one on epod-tsservice-dev.vgt.vito.be considered the "tsservice-openeo-dev" branch.

This means that if you send a request (with path prefix /v1.0 etc) to epod-tsservice-[12].vgt.vito.be:8080, you target a prod tsservice, whereas if you send one to epod-tsservice-dev.vgt.vito.be:8080, you target a dev tsservice.

In the Traefik case however, Traefik is responsible for both dev and prod/does not reside on epod-openeo-x machines. Therefore a `PathPrefix` matcher is not enough to distinguish between dev and prod so I've added `Host` matchers as well, and downstream proxies have to include the `Host` header.

**Comment by bossie** [comment]
tsservice request to Traefik with `Host` header works:

```
curl --location 'http://traefik01.vgt.vito.be/v1.0/ts/TERRASCOPE_S2_NDVI_V2/geometry?startDate=2017-01-01&endDate=2018-01-01' \
--header 'Content-Type: application/json' \
--header 'Host: epod-openeo-dev.vgt.vito.be' \
--data '{
  "type": "Polygon",
  "coordinates": [
    [
      [
        4.6770629938691854,
        50.82172692290532
      ],
      [
        4.6550903376191854,
        50.80697613242405
      ],
      [
        4.6866760309785604,
        50.797429020705295
      ],
      [
        4.7196350153535604,
        50.795692972629176
      ],
      [
        4.7402343805879354,
        50.81738893871384
      ],
      [
        4.6770629938691854,
        50.82172692290532
      ]
    ]
  ]
}'
```

**Comment by bossie** [comment]
F5 proxy update (point to Traefik cluster, pass `Host` header):
- must be handled by INF (Wim Mariën),
- at the same time as the promotion of OpenEO from dev to prod.

**Comment by bossie** [comment]
As discussed, will configure current nginx instances statically to route to Traefik instead of tsservice directly. This allows us to try this setup without relying on INF, and an easier roll back should it not work immediately.

Updating F5 and removing nginx from this chain of proxies can be done afterwards.

**Comment by bossie** [comment]
This static nginx config [works](http://epod-openeo-dev.vgt.vito.be:8080/internal/health) on epod-openeo-dev (disable gatekeeper and put it in /etc/nginx/conf.d/services_gatekeeper.conf):

```
     upstream traefik-dev {
          server traefik01-dev.vgt.vito.be:80;
          server traefik02-dev.vgt.vito.be:80;
      }

  server {
      listen 8080;

          location /internal {
              proxy_pass              http://traefik-dev;
              proxy_set_header        Host epod-openeo-dev.vgt.vito.be;
              proxy_connect_timeout   600;
              proxy_send_timeout      600;
              proxy_read_timeout      600;
              send_timeout      600;
          }
          location /v1.0 {
              proxy_pass              http://traefik-dev;
              proxy_set_header        Host epod-openeo-dev.vgt.vito.be;
              proxy_connect_timeout   600;
              proxy_send_timeout      600;
              proxy_read_timeout      600;
              send_timeout      600;
          }
          location /v1.1 {
              proxy_pass              http://traefik-dev;
              proxy_set_header        Host epod-openeo-dev.vgt.vito.be;
              proxy_connect_timeout   600;
              proxy_send_timeout      600;
              proxy_read_timeout      600;
              send_timeout      600;
          }
          location /apidocs {
              proxy_pass              http://traefik-dev;
              proxy_set_header        Host epod-openeo-dev.vgt.vito.be;
              proxy_connect_timeout   600;
              proxy_send_timeout      600;
              proxy_read_timeout      600;
              send_timeout      600;
          }
  }
```

**Comment by bossie** [comment]
nginx/gatekeeper on epod-openeo-1 got the same treatment; promoted dev to prod and embedded tsservice is now up with fixes for #356 and #357.

Tested these tsservice endpoints:
https://proba-v-mep.esa.int/api/timeseries/apidocs/
https://proba-v-mep.esa.int/api/timeseries/internal/health
https://proba-v-mep.esa.int/api/timeseries/v1.0/ts
https://proba-v-mep.esa.int/api/timeseries/v1.0/ts/S2_FAPAR_FILE/geometry?startDate=2017-01-01&endDate=2018-01-01

and some OpenEO tests as well.

**Comment by bossie** [comment]
Added follow up ticket #364 to remove workaround.
